### PR TITLE
docs: manual: point to `MutableArithmetics` in the Performance tips

### PR DESCRIPTION
**Pull request by nsajko**
Updates #31342
Updates #41792
Updates #50942

### CODE REVIEW

File: doc/src/manual/performance-tips.md [hunk 1]
```diff
@@ -1035,6 +1035,23 @@ some judgment may be required. However, for "vectorized" (element-wise) function
...
 (see the [dot syntax for vectorizing functions](@ref man-vectorized)).
 
+## [Use `MutableArithmetics` for more control over allocation for mutable arithmetic types](@id man-perftips-mutablearithmetics)
+
+Some [`Number`](@ref) subtypes, such as [`BigInt`](@ref) or [`BigFloat`](@ref), may
+be implemented as [`mutable struct`](@ref) types, or they may have mutable
+components. The arithmetic interfaces in Julia `Base` usually opt for convenience
+over efficiency in such cases, so using them in a naive manner may result in
+suboptimal performance. The abstractions of the
+[`MutableArithmetics`](https://juliahub.com/ui/Packages/MutableArithmetics/EoEec/)
```
Comment (brenhinkeller) [inline doc/src/manual/performance-tips.md:1045]: The `EoEec/` at the end of the URL seems to be superfluous. I'd probably also link the repo directly (`https://github.com/jump-dev/MutableArithmetics.jl`) but that's more a matter of taste

File: doc/src/manual/performance-tips.md [hunk 2]
```diff
@@ -1035,6 +1035,23 @@ some judgment may be required. However, for "vectorized" (element-wise) function
...
+package, on the other hand, make it possible to exploit the mutability of such types
+for writing fast code that allocates only as much as necessary. `MutableArithmetics`
+also makes it possible to copy values of mutable arithmetic types explicitly when
+necessary. `MutableArithmetics` is a user package and is not affiliated with the
+Julia project. The interfaces defined by `MutableArithmetics` may be implemented in
+`MutableArithmetics` itself, in some third-party package, or the relevant
+implementation may still be missing; depending on the specific mutable arithmetic
+type.
```
Comment (brenhinkeller) [inline doc/src/manual/performance-tips.md:1053]: This last sentence might be a bit more detail than needed; i'd probably delete it personally

**Review by brenhinkeller — APPROVED** [review]

**Comment by KristofferC** [comment]
Imo referencing an external package in the official manual that basically does "hackery" with the internals of some number types is not a good idea. If this was intended to be supported there would be a bunch of these utilities provided by Base but the official stance has always been that numbers are immutable. So with this PR there is kind of an inconsistency with that.

**Comment by nsajko** [comment]
Hmm, OK, interesting.

Some observations or opinions:

* It is precisely this official stance (of immutability of `Number`s) that makes `MutableArithmetics` necessary from a performance perspective in the ecosystem. So why not point to MA from the Performance tips?

* It seems to me like the existence of MA justifies this official stance and is inseparable from it. If there was no MA, surely this functionality would *have* to go into `Base`, or there would have to be another such package?

* I actually always assumed that MA, or some stripped-down version of MA, will enter `Base` at some point, not just because it would be useful to users, but mainly because it would be useful within `Base`. `Base` already has **internal** mutating functions that are used in some places for performance reasons, and MA would provide nicer interfaces for using those from within `Base`. My point being that I always assumed the existence of MA is kind of inevitable and necessary.

* MA is already used by many registered packages, so is there really a point to hiding it from other users?

**Comment by ViralBShah** [comment]
Perhaps we can slightly weaker the suggestion - and update the docs to point out that Number is immutable for good reason, but that some applications would benefit from MutableArithmetics?

I suppose we need to find the right balance between hiding it and recommending it. Often people take the performance tips in the julia manual to heart and start implementing them without careful consideration - because it ends up feeling like an official recommendation.

**Comment by ViralBShah** [comment]
Also apologies for the merge without adequate discussion. If necessary, we can revert and reopen the PR to get more views on board.

**Pull request merged**